Wip add ceph

oms4suse opened the pull request:
### What does this PR do?

Adds ceph support fro salt.

Add / create / import keyrings, create mon demons, create osd's, create rgw, create mds and the reverse of these options.

Test for ceph quorum.
### Tests written?

No

oms4suse committed:
salt.modules.ceph: Import execution module for ceph.

This ceph module depend upon a ceph configuration library python_ceph_cfg.

    https://github.com/oms4suse/python-ceph-cfg

This code was originally developed in the sesceph module:

    https://github.com/oms4suse/sesceph

It contains contributions from:

    Alex Lau (AvengerMoJo) <AvengerMoJo@gmail.com>
    Jan Fajerski <jfajerski@suse.com>

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Import execution module for ceph.

This ceph state depend upon a ceph configuration library python_ceph_cfg.

    https://github.com/oms4suse/python-ceph-cfg

This code was originally developed in the sesceph module:

    https://github.com/oms4suse/sesceph

Signed-off-by: Owen Synge <osynge@suse.com>

rallytime commented on salt/modules/ceph.py:
```diff
@@ -0,0 +1,654 @@
+import logging
```
The top of the file needs a file encoding line: `# -*- coding: utf-8 -*-`

Can you also add a doc string at the top of this file describing the new module? Be sure to include a `.. versionadded:: Carbon` directive in there so people know when this was added.

oms4suse replied:
I think this is now resolved.

rallytime commented on salt/modules/ceph.py:
```diff
@@ -0,0 +1,654 @@
+
+
+def osd_discover():
+    """
```
Can you change all of the `"""` references to single quotes like you have above?

rallytime commented on salt/modules/ceph.py:
```diff
@@ -0,0 +1,654 @@
+def zap(dev = None, **kwargs):
+    """
+    Destroy the partition table and content of a given disk.
+    """
```
CLI example?

rallytime commented on salt/modules/ceph.py:
```diff
@@ -0,0 +1,654 @@
+def ceph_version():
+    """
+    Get the version of ceph installed
+    """
```
CLI Example here too.

thatch45 commented:
The failing tests are nothing to worry about, they are unrelated to your PR, the only thing to fix up is the lint:
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3541/violations/file/salt/modules/ceph.py/
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3541/violations/file/salt/states/ceph.py/

The tricky bit is the `Incompatible python3 code found` lines, fix them by making the first import:
`from __future__ import absolute_import`

We get these lint errors fixed and we are ready to merge! Thanks @oms4suse !

rallytime commented on salt/states/ceph.py:
```diff
@@ -0,0 +1,59 @@
+import logging
```
This state file needs some initial docs and a file encoding like as well. :)

oms4suse replied:
I think this is now resolved.

rallytime commented on salt/states/ceph.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def quorum(name, **kwargs):
+    """
```
Single quotes here too, please.

rallytime commented on salt/states/ceph.py:
```diff
@@ -0,0 +1,59 @@
+
+    This state is needed to allow the cluster to function.
+
+    Example usage in sls file:
```
Can you add a `.. code-block:: yaml` directive here, and then adjust the spacing for the example yaml file below?

rallytime commented on salt/states/ceph.py:
```diff
@@ -0,0 +1,59 @@
+    if paramters is None:
+        return _error(name, "Invalid paramters:%s")
+
+    if __opts__['test']:
```
Can you refactor this `test` part here so that if you're passing test=true, you still get a list of changes? We also don't want to just return `None` here if no changes are expected. If test=true is passed in, but the state will not change, then we don't want to return None (indicating a state change).

There are some docs about state returns here, for more info: https://docs.saltstack.com/en/latest/ref/states/writing.html#return-data

I also don't see anywhere where you're checking for changes or using the `_changed` function.

oms4suse replied:
This method does not change anything.

thatch45 replied:
The problem here is not that no changes are being made, it is that no changes are checked for. Before we return the test=True results we want to first check and see if changes need to be made. If changes need to be made then we return result=None, but if changes do not need to be made then we return result=True, indicating that running this state will not change anything, and thus making sure it is idempotent and able to accurately convey the data from the idempotent operation.

cachedout replied:
@oms4suse Just waiting on this final issue to be addressed before we get this in. Thanks!

thatch45 replied:
@oms4suse please let us know how we can help! We all really want to get this in!

rallytime commented:
Thanks for this awesome start! I had a couple of comments, which are mostly simple doc changes. There are also a couple of pylint errors that need to be cleaned up: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3541/violations/

rallytime commented:
Oops - looks like Tom and I were reviewing this at the same time. And most of my comments were covered by lint anyway...sorry about that.

oms4suse committed:
salt.modules.ceph: Define the text as UTF 8

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Define the text as UTF 8

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: Use triple ' in block comments.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Use triple ' in block comments.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: Add CLI examples

Add CLI paramters and CLI exampls to the methods of the module ceph

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Mark up the YAML.

The state method should have the YAML marked up.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.moduels.ceph: Add module metadata

Salt has a rich metadata format for all modules.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Add module metadata

Salt has a rich metadata format for all state modules.

Also improved the online help.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: shortened line

Trying to pass pylint-3.4

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: bad-whitespace

Trying to pass pylint-3.4

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: Fixed unused paramter

Trying to pass pylint-3.4

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: shortened line

Trying to pass pylint-3.4

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse commented:
Hope this version will pass some more tests.

I will be going on vacation on Wednesday for 10 days, just in case you think I have disappeared.

oms4suse committed:
salt.modules.ceph: absolute_import

Avoid unexpected behaviour with imports.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: absolute_import

Avoid unexpected behaviour with imports.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Rename exception.

pylint3 does not like single character variables.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: Remove tabs

Seems some tabs got into the code base.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Add docstring to utils

Non exported utils also need docstring to pass pylint3

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.modules.ceph: Use format not substitution

We should use new style format in salt code.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse committed:
salt.states.ceph: Remove HAS_CEPH_CFG

pylint does not like unused imports so even though we will re add this later, we
remove this now to get the code upstreamed.

Signed-off-by: Owen Synge <osynge@suse.com>

oms4suse commented:
@rallytime  @thatch45 thatch pylint is now passing. :)

oms4suse committed:
salt.modules.ceph: Fix indentation : keyring_present

Make indentation the same as other methods.

Signed-off-by: Owen Synge <osynge@suse.com>

thatch45 commented:
Thanks for the cleanup @oms4suse !!
The failing tests are not related so I think that this is likely good to merge!
@rallytime please let me know if you feel otherwise

thatch45 commented:
Yes, I think that the only concern we have is the state module test=True facility. I am willing to merge this in before that is fixed, but we should make sure that fix is something we can reasonably do

cachedout commented:
I've gone ahead and merged this but let's make sure we follow-up here with test=True support. Thanks for everyhting @oms4suse !

thatch45 commented:
Good call @cachedout ! This is not a serious issue, we are just striving for perfection! Thanks again @oms4suse , we are all very excited about this.

oms4suse commented:
@rallytime  @thatch45 Any thing I can do to get this in, I have not seen any updates while I was on holiday?

oms4suse commented:
@rallytime @thatch45 Ops just seen this is merged. Thanks. I will start working on a better state module with better testing to make @cachedout  happy :)

thatch45 commented:
Thanks @oms4suse ! Let us know if you need anything on the next step
